File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (root == nullptr) return nullptr;
        queue<Node*> q;
        q.push(root);
        while (!q.empty()) {
            int size = q.size();
            Node* tmp = nullptr;
            for (int i = 0; i < size; i++) {
                Node* n = q.front();
                q.pop();
                n->next = tmp;
                tmp = n;
                if (n->right != nullptr) q.push(n->right);
                if (n->left != nullptr) q.push(n->left);
            }
        }
        return root;
    }
};

File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if (root == null) return root;
        connect(root.left, root.right);
        return root;
    }
    
    void connect(Node leftRoot, Node rightRoot) {
        if (leftRoot == null || rightRoot == null) return;
        
        leftRoot.next = rightRoot;
        connect(leftRoot.left, leftRoot.right);
        connect(leftRoot.right, rightRoot.left);
        connect(rightRoot.left, rightRoot.right);
    }
}